### PR TITLE
str_angles and scale_units logic for quiver

### PR DESCRIPTION
**Pull request by neok-m4700**
**issue 1**
https://github.com/matplotlib/matplotlib/commit/d705dbd2488b235ba173515ff29b586f4c6d74cd is introducing issues when using `angles='xy'`, especially at l.660:
```python
if str_angles and (angles == 'xy'):
    theta = angles
```
, comparing angles (now an array) to a string is throwing a warning
```python
FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
```

**issue 2**
using `scale_units='xy'` together with an angles array throws an
```python
UnboundLocalError: local variable 'lengths' referenced before assignment
```
where it shouldn't

**reproduce the problem**
```python
import numpy as np
import matplotlib.pyplot as plt

# logic issue (angles: str is overwritten with an array), theta is wrong
_, ax = plt.subplots(subplot_kw=dict(aspect='equal'))
ax.quiver(0, 0, 1, 1, angles='xy', scale_units='xy', scale=1)
ax.set_xlim(0, 1.1); ax.set_ylim(0, 1.1); ax.grid()

# scale_units='xy' + angles array issue
_, ax = plt.subplots()
X, Y = np.meshgrid(np.arange(15), np.arange(10))
U = V = np.ones_like(X)
phi = (np.random.rand(15, 10) - .5) * 150
ax.quiver(X, Y, U, V, angles=phi, scale_units='xy')
```

this patch fixes both issues

### CODE REVIEW

File: lib/matplotlib/quiver.py [hunk 1]
```diff
@@ -626,19 +626,19 @@ def _angles_lengths(self, U, V, eps=1):
 
     def _make_verts(self, U, V, angles):
         uv = (U + V * 1j)
-        str_angles = isinstance(angles, six.string_types)
-        if str_angles and (angles == 'xy' and self.scale_units == 'xy'):
+        str_angles = angles if isinstance(angles, six.string_types) else ''
```
Comment (QuLogic) [inline lib/matplotlib/quiver.py]: I'm not seeing why this change is necessary.

Comment (neok-m4700) [inline lib/matplotlib/quiver.py]: Build mpl on latest commit 1144e8c0b89434888a15b1148e1aea21db9878d4 on branch master

Run
```python
_, ax = plt.subplots(subplot_kw=dict(aspect='equal'))
ax.quiver(0, 0, 1, 1, angles='xy', scale_units='xy', scale=1)
ax.set_xlim(0, 1.1); ax.set_ylim(0, 1.1); ax.grid()
```
[before.pdf](https://github.com/matplotlib/matplotlib/files/1033118/before.pdf)
[after.pdf](https://github.com/matplotlib/matplotlib/files/1033117/after.pdf)

this block l. 668 is wrong:
```python
        if str_angles and (angles == 'xy'):
            theta = angles
        elif str_angles and (angles == 'uv'):
            theta = np.angle(uv)
```
since variable angles is overridden by `angles, lengths = self._angles_lengths(U, V, eps=1)`

Comment (QuLogic) [inline lib/matplotlib/quiver.py]: Ah, okay. In that case, `if str_angles and str_angles == 'some string':` is redundant now; it can just be `if str_angles == 'some string':`.

Please add a test as well; something similar to your example will probably work.

Comment (neok-m4700) [inline lib/matplotlib/quiver.py]: I guess not, str_angle is here to avoid the `FutureWarning: elementwise comparison failed` cited above and was introduced in https://github.com/matplotlib/matplotlib/pull/6971

I tried to stick with the existant ...

Comment (QuLogic) [inline lib/matplotlib/quiver.py]: But this line has changed it so that `str_angles` is _always_ a string.

Comment (neok-m4700) [inline lib/matplotlib/quiver.py]: yes, but the empty string evaluates to False (as in `print('foo' if '' else 'bar')`), thus short-circuiting the angles == 'xy' test, which can eventually throw a warning.

Comment (QuLogic) [inline lib/matplotlib/quiver.py]: But you've replaced all the `angles ==` with `str_angles ==`.

Comment (neok-m4700) [inline lib/matplotlib/quiver.py]: ha yes I see it now, stupid me it's getting late here ...

File: lib/matplotlib/quiver.py [hunk 2]
```diff
@@ -626,19 +626,19 @@ def _angles_lengths(self, U, V, eps=1):
...
             angles, lengths = self._angles_lengths(U, V, eps=1)
-        elif str_angles and (angles == 'xy' or self.scale_units == 'xy'):
+        elif str_angles == 'xy' or self.scale_units == 'xy':
             # Calculate eps based on the extents of the plot
             # so that we don't end up with roundoff error from
             # adding a small number to a large.
             eps = np.abs(self.ax.dataLim.extents).max() * 0.001
             angles, lengths = self._angles_lengths(U, V, eps=eps)
-        if self.scale_units == 'xy':
+        if str_angles and self.scale_units == 'xy':
```
Comment (dstansby) [inline lib/matplotlib/quiver.py]: Why does this extra bit need to go in - if str_angles is a string does this even make sense?

File: lib/matplotlib/tests/test_quiver.py [hunk 3]
```diff
@@ -155,6 +155,24 @@ def test_bad_masked_sizes():
...
+    ax.quiver(X, Y, U, V, angles=phi, scale_units='xy')
+
+
+@image_comparison(baseline_images=['quiver_xy'],
+                  extensions=['png'], remove_text=True)
+def test_quiver_xy():
+    # simple arrow pointing from SW to NE
+    fig, ax = plt.subplots(subplot_kw=dict(aspect='equal'))
+    ax.quiver(0, 0, 1, 1, angles='xy', scale_units='xy', scale=1)
+    ax.set_xlim(0, 1.1); ax.set_ylim(0, 1.1); ax.grid()
```
Comment (dstansby) [inline lib/matplotlib/tests/test_quiver.py]: Could these all go on a separate line without semicolons

File: lib/matplotlib/quiver.py (hunk 2 above)
Comment (neok-m4700) [inline lib/matplotlib/quiver.py]: this should avoid throwing issue 2 is str_angles == '', doesn't it ?

File: lib/matplotlib/tests/test_quiver.py (hunk 3 above)
Comment (neok-m4700) [inline lib/matplotlib/tests/test_quiver.py]: done. the issue with the test_image is mostly due to different version of fontconfig on my dev version, I will try to generate it using mpl from the repos ... thanks

File: lib/matplotlib/quiver.py (hunk 2 above)
Comment (QuLogic) [inline lib/matplotlib/quiver.py]: Maybe it should be `if str_angles == 'xy'` as well, to match the others? I'm not sure what the behaviour should be if you pass `angles = 'some garbage'` instead though.

Comment (QuLogic) [inline lib/matplotlib/quiver.py]: I guess the real question is whether this should apply to `str_angles == 'uv'` as well?

**Comment by QuLogic** [comment]
Once you've answered the last question, can you rebase and remove the multiple copies of the test image?

File: lib/matplotlib/quiver.py (hunk 2 above)
Comment (neok-m4700) [inline lib/matplotlib/quiver.py]: well, all the scalings except 'xy' are based on U and V, so I guess I it sufficient to leave it like that. throwing exception on `str_angles not in ('xy', 'uv')` seems to be an overkill since `a` and `theta` will fallback to using `U` and `V` values ...

Comment (QuLogic) [inline lib/matplotlib/quiver.py]: Appears that `angles='uv'` is the default, so the existing tests not being changed shows that it's probably fine.

**Comment by tacaswell** [comment]
:tada: Congratulations on your first merged PR into Matplotlib @neok-m4700 !  Hopefully we will hear from you again!

**Comment by neok-m4700** [comment]
Thanks @tacaswell , I guess there is a start to everything :)
